File: ACM/LanQiao/2172/2172.py
'''
Author: NEFU AB-IN
Date: 2023-05-24 12:47:50
FilePath: \LanQiao\2172\2172.py
LastEditTime: 2023-05-24 14:52:22
'''
# import
from sys import setrecursionlimit, stdin, stdout, exit
from collections import Counter, deque
from heapq import heapify, heappop, heappush, nlargest, nsmallest
from bisect import bisect_left, bisect_right
from datetime import datetime, timedelta
from string import ascii_lowercase, ascii_uppercase
from math import log, gcd, sqrt, fabs, ceil, floor


class sa:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __lt__(self, a):
        return self.x < a.x


# Final
N = int(1e5 + 10)
M = 20
INF = int(2e9)

# Define
setrecursionlimit(INF)
input = lambda: stdin.readline().rstrip("\r\n")  # Remove when Mutiple data
read = lambda: map(int, input().split())
LTN = lambda x: ord(x.upper()) - 65  # A -> 0
NTL = lambda x: ascii_uppercase[x]  # 0 -> A

# —————————————————————Division line ——————————————————————
a = [0] * N
dp = [[0] * M for _ in range(N)]
Log = [0] * N


def init():
    for j in range(M):
        i = 1
        while i + (1 << j) - 1 <= n:
            if j == 0:
                dp[i][j] = a[i]
            else:
                dp[i][j] = gcd(dp[i][j - 1], dp[i + (1 << (j - 1))][j - 1])
            i += 1
    Log[1] = 0
    for i in range(2, N):
        Log[i] = Log[i >> 1] + 1


def query(l, r):
    k = Log[r - l + 1]
    return gcd(dp[l][k], dp[r - (1 << k) + 1][k])


n, = read()
a[1:] = read()

init()
cnt = sum(i == 1 for i in a)
if cnt > 0:
    print(n - cnt)
    exit(0)

if query(1, n) != 1:
    print(-1)
    exit(0)

ans = INF

for i in range(1, n + 1):
    l, r = i, n
    while l < r:
        mid = (l + r) >> 1
        if query(i, mid) == 1:
            r = mid
        else:
            l = mid + 1
    if query(i, r) == 1:
        ans = min(ans, r - i)

print(ans + n - 1)

File: ACM/LanQiao/2172/2172.cpp
/*
 * @Author: NEFU AB-IN
 * @Date: 2023-05-24 12:55:42
 * @FilePath: \LanQiao\2172\2172.cpp
 * @LastEditTime: 2023-05-24 12:55:55
 */
#include <bits/stdc++.h>
using namespace std;
#define int long long
#undef int

#define SZ(X) ((int)(X).size())
#define ALL(X) (X).begin(), (X).end()
#define IOS                                                                                                            \
    ios::sync_with_stdio(false);                                                                                       \
    cin.tie(nullptr);                                                                                                  \
    cout.tie(nullptr)
#define DEBUG(X) cout << #X << ": " << X << '\n'
typedef pair<int, int> PII;

const int N = 1e5 + 10, INF = 0x3f3f3f3f;

const int N = 200010, M = 18;

int n, m;
int w[N], f[N][M], Log[N]; // 用来求log的

void init()
{
    for (int j = 0; j < M; j++)
    {
        for (int i = 1; i + (1 << j) - 1 <= n; i++)
        {
            if (!j)
                f[i][j] = w[i];
            else
                f[i][j] = max(f[i][j - 1], f[i + (1 << (j - 1))][j - 1]);
        }
    }
    Log[1] = 0;
    for (int i = 2; i <= n; i++)
        Log[i] = Log[i / 2] + 1;
}

int query(int l, int r)
{
    int k = Log[r - l + 1];
    return max(f[l][k], f[r - (1 << k) + 1][k]);
}

signed main()
{
    IOS;

    return 0;
}